### DISCUSSION TITLE
Build errors (MapboxGLGlyphManager related)

### DISCUSSION THREAD

**Discussion by MarkusMoss** [pelicanmapping/osgearth#2007] [Support Q&A]
Hi, I have tried  to build latest osgearth code  (try linking it  against latest OSG). 

For me, building osgearth (under latest ubuntu),  stops here with these  error messages:

"osgearth/src/osgEarth/MapboxGLGlyphManager.cpp:25:10: fatal error: glyphs.pb.h: Fila eller mappa finnes ikke " (file not found)

and

"osgearth/src/osgEarth/MapboxGLImageLayer.cpp:243:27: error: ‘MVTFeatureSource’ was not declared in this scope; did you mean ‘featureSource’?
  243 |             osg::ref_ptr< MVTFeatureSource > featureSource = new MVTFeatureSource();"

Do I need to install some "Mapbox" 3rdparty to get it built ? 

Thanks for info ! regards, Markus

**Comment by jasonbeverage** [comment]
You need to build with protocol buffer support by building against
protobuf.  It looks I didn't protect that code when protobuf wasn't present.

**Reply by nebsar** [reply]
I had the same problem and I commented out that lines while building.

**Reply by MarkusMoss** [reply]
@nebsar thanks, I also will outcomment these lines in order to get it built.

**Comment by jasonbeverage** [comment]
Is there a reason you can't build against protobuf?  Honestly I'd like to
just make it a required dependency.  We use vector tiles all the time which
requires protobuf so we always build with it on.

Jason

**Reply by nebsar** [reply]
There is no reason. Actually after that I built the latest version against protobuf. But this time I got error on MapboxGLImageLayer.cpp at line 236, since do not have MVT support. And I commented out the following lines:

if (type() == "vector-mbtiles")
        {

            URI uri(url(), context);

            osg::ref_ptr< MVTFeatureSource > featureSource = new MVTFeatureSource();
            featureSource->setReadOptions(options);
            featureSource->setURL(uri);
            featureSource->open();
            _featureSource = featureSource.get();

        }

**Comment by jasonbeverage** [comment]
You might need to rerun cmake.  If you just added protobuf as a dependency
it needs to regenerate the project and you should have support for loading
MVT.

**Reply by nebsar** [reply]
Ok I will try. Thanks.

**Reply by nebsar** [reply]
Hello Jason,

I tried to build osgEarth again. But the main problem is I do not have vector_tile.pb.h header. I think this is about MapBox Mapnik library. Is it right? How can I include that library?

**Comment by MarkusMoss** [comment]
Thanks for info and fast reply!

I will try nebsar's approach.
Seems to me that MapboxGLGlyphManager is under devolopment.
I havn't built osgearth during the last 3-4 years.., lots of new promising 3rdparty libs there, and the cmake-config overloaded with options :-)
 I have to dig a little bit into the changelog.

Markus

**Comment by jasonbeverage** [comment]
That file is generated from the  vetor_tile.proto file in the osgearth
source code.  On Linux you need to install the protobuf-compiler and
libprotobuf-dev like we do in our Github actions here
https://github.com/gwaldron/osgearth/blob/master/.github/workflows/linux.yml#L31
Once you have that re-rerun cmake and the protoc compiler will generate
that header for you from the proto file.

Jason

**Reply by nebsar** [reply]
Thank you Jason,
What directory should I show in the cmake in the picture? "src/protobuf/google/"? Cmake does not seems to find protobuf source code automatically. Sorry for the stupid question.

![image](https://user-images.githubusercontent.com/5169314/156838933-f913ee6e-4de7-4f77-abd2-1554a240cfa2.png)

**Comment by gwaldron** [comment]
I recently pushed a fix that lets you build without protobuf, so that's another option.

Otherwise, you can leave "Protobuf_SRC_ROOT_FOLDER" unset.

**Reply by nebsar** [reply]
Thanks Glenn.

**Reply by MarkusMoss** [reply]
Thanks!

**Reply by marcoma9023** [reply]
I tried to compile, keeping the Protobuf_SRC_ROOT_FOLDER empty, but at the end of the compilation it gives me compile error due to #include "glyphs.pb.h" ... so the compilation of OSGEarth fails ...

I also tried to pass the folder containing protoc.exe as src folder of protobuf, but nothing ....

What am I doing wrong?

**Comment by nebsar** [comment]
I managed to build the latest version with protobuf at last 😁, with MVT support. Protobuf_SRC_ROOT_FOLDER should indicate the folder that protoc.exe exists.

**Reply by nebsar** [reply]
The first time I build the osgEarth with protobuf, but now, it gives run time shader errors ☹

**Comment by gwaldron** [comment]
Converted to issue #2042.

**Comment by jasonbeverage** [comment]
What operating system are you using?

You shouldn't need to set that manually usually, if you are using Linux it
should just find it if you installed protobuf via apt get.  On Windows I'd
recommend building using vcpkg to get your dependencies as described here:
http://docs.osgearth.org/en/latest/build.html

Jason
